### PR TITLE
Package manager pnpm

### PR DESCRIPTION
**Pull request by dishuostec**
# ↪️ Pull Request

Try to add [pnpm](https://github.com/pnpm/pnpm) suport

#3408

### CODE REVIEW

File: packages/core/package-manager/src/installPackage.js [hunk 1]
```diff
@@ -128,13 +129,19 @@ async function determinePackageInstaller(
...
     'package-lock.json',
+    'pnpm-lock.yaml',
   ]);
   let hasYarn = await Yarn.exists();
 
   // If Yarn isn't available, or there is a package-lock.json file, use npm.
   let configName = configFile && path.basename(configFile);
-  if (!hasYarn || configName === 'package-lock.json') {
-    return new Npm();
+  if (!hasYarn || configName) {
```
Comment (101arrowz) [inline packages/core/package-manager/src/installPackage.js:138]: If the Yarn binary doesn't exist and no lockfile is present, NPM should be used by default. However, if there is no lockfile, this new logic will cause Yarn to be used even when the binary doesn't exist.

Comment (dishuostec) [reply packages/core/package-manager/src/installPackage.js:138]: > could you add some integration tests? You can take inspiration from how we did it for yarn 2



> If the Yarn binary doesn't exist and no lockfile is present, NPM should be used by default. However, if there is no lockfile, this new logic will cause Yarn to be used even when the binary doesn't exist.

Yes, I made a mistake.

**Comment by dishuostec** [comment]
> could you add some integration tests? You can take inspiration from how we did it for yarn 2

I couldn't find which part is the test for yarn2, do you mean `@parcel/integration-test`?

**Comment by DeMoorJasper** [comment]
@dishuostec yes that's the package with all our integration tests.

For inspiration I meant https://github.com/parcel-bundler/parcel/blob/v2/packages/core/integration-tests/test/pnp.js which is part of our test suite

**Comment by mischnic** [comment]
(The PnP tests are only testing the resolver, not the autoinstall part)

**Comment by dishuostec** [comment]
Pnpm use symlinked node_modules structure. If a package is dependent on by several package, then the package will be resolved as different position. It will be bundle multiple times.

I created new pr https://github.com/parcel-bundler/parcel/pull/4939.

**Comment by aminya** [comment]
Really looking forward to this. Having to manage many packages, the time and storage that pnpm saves me is a big advantage. However, the fact that Parcel does not work with pnpm is a bummer!

**Pull request closed without merging**